File: Wybrane_Elementy_Praktyki_Projektowania_Oprogramowania/Lista4/Zad6.js
function Tree(val, left, right) {
    this.left = left;
    this.right = right;
    this.val = val;
}

/*
Tree.prototype[Symbol.iterator] = function* () {
    yield this.val;
    if (this.left) yield* this.left;
    if (this.right) yield* this.right;
}
*/

var root = new Tree(1, new Tree(2, new Tree(3)), new Tree(4));

Tree.prototype[Symbol.iterator] = function* () {
    let queue = [this];

    while (queue.length > 0) {
        let node = queue[0];
        queue.splice(0, 1);
        if (node.right) {
            queue.splice(queue.length, 0, node.right);
            //queue.splice(0, 0, node.right);
        } 
        if (node.left){
            queue.splice(queue.length, 0, node.left);
            //queue.splice(0, 0, node.left);
        } 
        yield node.val;
    }
};

for (var e of root) {
    console.log(e);
}

File: Programowanie_Pod_Windows(.NET)/Lista3/Zad6.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Zad6
{
    public class BinaryTreeNode<T>
    {
        public BinaryTreeNode<T> Left { get; set; }
        public BinaryTreeNode<T> Right { get; set; }
        public T value { get; set; }

        public IEnumerable<T> DFS()
        {
            yield return this.value;
            if (this.Left != null)
            {
                foreach (T t in this.Left.DFS())
                {
                    yield return t;
                }

            }
            if (this.Right != null)
            {
                foreach (T t in this.Right.DFS())
                {
                    yield return t;
                }

            }
        }

        public IEnumerable<T> BFS()
        {
            Queue<BinaryTreeNode<T>> queue = new Queue<BinaryTreeNode<T>>();
            queue.Enqueue(this);
            while(queue.Count > 0)
            {
                BinaryTreeNode<T> t = queue.Dequeue();
                yield return t.value;
                if(t.Left != null)
                {
                    queue.Enqueue(t.Left);
                }
                if(t.Right != null)
                {
                    queue.Enqueue(t.Right);
                }
            }
        }
    }
    internal class Program
    {
        static void Main(string[] args)
        {
            BinaryTreeNode<int> root = new BinaryTreeNode<int>
            {
                Left = new BinaryTreeNode<int>
                {
                    Left = new BinaryTreeNode<int>
                    {
                        Left = null,
                        value = 6,
                        Right = null
                    },
                    value = 8,
                    Right = new BinaryTreeNode<int>
                    {
                        Right = null,
                        value = 9,
                        Left = null
                    }
                },
                value = 12,
                Right = new BinaryTreeNode<int>
                {
                    Left = new BinaryTreeNode<int>
                    {
                        Left = null,
                        value = 16,
                        Right = null
                    },
                    value = 18,
                    Right = new BinaryTreeNode<int>
                    {
                        Right = null,
                        value = 24,
                        Left = null
                    }
                }
            };

            Console.WriteLine("BFS");
            foreach(int i in root.BFS())
            {
                Console.WriteLine(i);
            }

            Console.WriteLine("DFS");
            foreach (int i in root.DFS())
            {
                Console.WriteLine(i);
            }
            Console.ReadLine();
        }
    }
}
